src/paulie/application/pauli_compiler_optimal.py

```diff
@@ -0,0 +1,612 @@
+    """
+    return a @ b
+
+def _commutes(a: PauliString, b: PauliString) -> bool:
+    """
+    Check on commutes two PauliStrings
+    """
+    return a | b
+
+def _left_part(p: PauliString, k: int) -> PauliString:
+    """
+    Get left substring of PauliString
+    """
+    return p.get_substring(0, k)
+
+def _right_part(p: PauliString, k: int) -> PauliString:
+    """
+    Get right substring of PauliString
+    """
+    return p.get_substring(k, len(p) - k)
+
+def _ad_apply(A: PauliString, B: PauliString | None) -> PauliString | None:
+    """
+    Compute the adjoint map ad_A(B) = [A,B]
+    Returns None if the commutator is zero (i.e., if A and B commute)
+    Otherwise returns a PauliString proportional to the commutator
+    """
+    if B is None:
+        return None
+    return None if _commutes(A, B) else _multiply(A, B)
+
+def _nested_commutator_result(G: list[PauliString]) -> PauliString | None:
+    """
+    Get nested commutator
+    """
+    if not G:
+        return None
+    cur: PauliString | None = G[0]
+    for A in G[1:]:
+        cur = _ad_apply(A, cur)
+        if cur is None:
+            return None
+    return cur
+
+def _sequence_to_paulie_orientation(G: list[PauliString]) -> list[PauliString]:
+    """
+    PauLie nested_adjoint iterates operators in reversed order.
+    Our G is [base, A1, A2, ..., Am] where main.py applies A1 then A2 ... then Am.
+    To make nested_adjoint(seq[:-1], seq[-1]) apply in the same order, we must
+    return operators reversed: [Am, ..., A2, A1, base].
+    """
+    if not G:
+        return []
+    base = G[0]
+    ops = G[1:]
+    ops_rev = list(reversed(ops))
+    return ops_rev + [base]
+
+
+def left_a_minimal(k: int) -> list[PauliString]:
+    """Return minimal adjoint-universal set on the left: {Xi, Zi} plus Z...Z."""
+    a_ops: list[PauliString] = []
+    for i in range(k):
+        a_ops.append(_single(k, i, "X"))
+        a_ops.append(_single(k, i, "Z"))
+    zall = get_pauli_string("Z" * k)
+    a_ops.append(zall)
+    return a_ops
+
+def choose_u_for_b(k: int) -> PauliString:
+    """Choose the left tag operator used for coupling to right-side B's."""
+    return _single(k, 0, "X")
+
+def _all_left_paulis(k: int) -> list[PauliString]:
+    """ Enumerate all non-identity Pauli strings of length k"""
+    out: list[PauliString] = []
+    labels = ["I", "X", "Y", "Z"]
+    def rec(i: int, cur: list[str]):
+        if i == k:
+            s = "".join(cur)
+            if set(s) != {"I"}:
+                out.append(get_pauli_string(s))
+            return
+        for L in labels:
+            cur.append(L)
+            rec(i + 1, cur)
+            cur.pop()
+    rec(0, [])
+    return out
+
+
+@dataclass
+class SubsystemCompilerConfig:
+    """Config of subsystem compiler"""
+    k_left: int
+    n_total: int
+
+class SubsystemCompiler:
+    """
+    subsystem compiler
+    """
+    def __init__(self, cfg: SubsystemCompilerConfig):
+        """
+        constructor
```

Comment:
This comment is WET, we should remove or replace with a better one. There are also many other comments like this below that we should apply the same to.